esphome/components/http_request/http_request_idf.cpp

```diff
@@ -0,0 +1,139 @@
+#include "http_request_idf.h"
+
+#ifdef USE_ESP_IDF
+
+#include "esphome/components/network/util.h"
+#include "esphome/core/application.h"
+#include "esphome/core/defines.h"
+#include "esphome/core/log.h"
+
+namespace esphome {
+namespace http_request {
+
+static const char *const TAG = "http_request.idf";
+
+esp_err_t http_event_handler(esp_http_client_event_t *evt) {
+  App.feed_wdt();
+  switch (evt->event_id) {
+    case HTTP_EVENT_ERROR:
+      ESP_LOGE(TAG, "HTTP_EVENT_ERROR");
+      break;
+    case HTTP_EVENT_ON_CONNECTED:
+      ESP_LOGV(TAG, "HTTP_EVENT_ON_CONNECTED");
+      break;
+    case HTTP_EVENT_HEADER_SENT:
+      ESP_LOGV(TAG, "HTTP_EVENT_HEADER_SENT");
+      break;
+    case HTTP_EVENT_ON_HEADER:
+      ESP_LOGV(TAG, "HTTP_EVENT_ON_HEADER, key=%s, value=%s", evt->header_key, evt->header_value);
+      break;
+    case HTTP_EVENT_ON_DATA:
+      ESP_LOGV(TAG, "HTTP_EVENT_ON_DATA, len=%d", evt->data_len);
+      /*
+       *  Check for chunked encoding is added as the URL for chunked encoding used in this example returns binary data.
+       *  However, event handler can also be used in case chunked encoding is used.
+       */
+      if (!esp_http_client_is_chunked_response(evt->client)) {
+        if (global_http_request->get_capture_response()) {
+          std::vector<char> &buffer = *reinterpret_cast<std::vector<char> *>(evt->user_data);
+          buffer.insert(buffer.end(), reinterpret_cast<char *>(evt->data),
+                        reinterpret_cast<char *>(evt->data) + evt->data_len);
+        }
+      }
+
+      break;
+
+    case HTTP_EVENT_ON_FINISH:
+      ESP_LOGV(TAG, "HTTP_EVENT_ON_FINISH");
+      break;
+    case HTTP_EVENT_DISCONNECTED:
+      ESP_LOGV(TAG, "HTTP_EVENT_DISCONNECTED");
+      break;
+  }
+  return ESP_OK;
+}
+
+void HttpRequestIDF::set_url(std::string url) { this->url_ = std::move(url); }
+
+HttpResponse HttpRequestIDF::send(bool capture_response) {
+  if (!network::is_connected()) {
+    this->status_set_warning();
+    ESP_LOGE(TAG, "HTTP Request failed; Not connected to network");
+    return {-1, 0, {}};
+  }
+
+  std::vector<char> buffer;
+
+  esp_http_client_method_t method;
+  if (this->method_ == "GET") {
+    method = HTTP_METHOD_GET;
+  } else if (this->method_ == "POST") {
+    method = HTTP_METHOD_POST;
+  } else if (this->method_ == "PUT") {
+    method = HTTP_METHOD_PUT;
+  } else if (this->method_ == "DELETE") {
+    method = HTTP_METHOD_DELETE;
+  } else if (this->method_ == "PATCH") {
+    method = HTTP_METHOD_PATCH;
+  } else {
+    this->status_set_warning();
+    ESP_LOGE(TAG, "HTTP Request failed; Unsupported method");
+    return {-1, 0, {}};
+  }
+
+  esp_http_client_config_t config = {};
+
+  config.url = this->url_.c_str();
+  config.method = method;
+  config.timeout_ms = this->timeout_;
+  config.disable_auto_redirect = !this->follow_redirects_;
+  config.max_redirection_count = this->redirect_limit_;
+  config.user_data = (void *) &buffer;
```

Comment:
```suggestion
  config.user_data = nullptr;
```

Actually, reading this again.

`&buffer` is the address of `buffer` on the stack.
It has nothing to do with the underlying memory allocated,  for that you want `.data()`.

However, `buffer` never actually allocates;  you never use it.
Explosions all the way down.

I recommend you leave this as `nullptr`.
You could probably use `evt->data` directly in `http_event_handler`, depending on it's lifetime.
No `std::vector` required.